#3 로그인 기능 구현

hamryt opened the pull request:
session을 사용하여 로그인을 구현함

hamryt committed:
#3 로그인 기능 구현
session을 사용하여 로그인을 구현함

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/controller/UserController.java:
```diff
@@ -17,12 +30,73 @@
     }
 
+    @PostMapping("login")
+    public ResponseEntity<LoginResponse> login(
```
여기서는 `LoginResponse`만 리턴해도 정상적으로 동작합니다~

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/controller/UserController.java:
```diff
@@ -17,12 +30,73 @@
+        UserDto userInfo = userService.loginUser(email, password);
+        LoginResponse loginResponse;
+
+        if (userInfo == null) {
+            loginResponse = LoginResponse.FAIL;
+            responseEntity = new ResponseEntity<LoginResponse>(loginResponse,
+                HttpStatus.UNAUTHORIZED);
+        } else if (UserDto.Status.DEFAULT.equals(userInfo.getStatus())) {
+            loginResponse = LoginResponse.success(userInfo);
+            SessionUtil.setLoginUserEmail(session, email);
+            responseEntity = new ResponseEntity<LoginResponse>(loginResponse, HttpStatus.OK);
+        } else {
+            log.error("Login Error");
+            throw new RuntimeException("Login Error");
+        }
```
이 로직들은 비즈니스 로직으로 보입니다~ `LoginService`를 만들어 그곳으로 책임을 분리하는게 낫지 않을까요?

hamryt replied:
넵 분리하겠습니다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/controller/UserController.java:
```diff
@@ -17,12 +30,73 @@
+
+        return responseEntity;
+    }
+    //------------------------ request 객체 ---------------------------
```
이런것들은 굳이 써주시지 않아도 되긴 합니다~

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/util/SessionUtil.java:
```diff
@@ -0,0 +1,15 @@
+
+import javax.servlet.http.HttpSession;
+
+public class SessionUtil {
```
이 클래스는 인증을 하는 비즈니스 로직을 담고있습니다~ 유틸클래스가 아니라 bean으로 선언하는게 좋지 않을까요? 이름도 적절하게 바꿔보고요

hamryt committed:
로그인 기능 구현 코드 재배치

UserController에 있는 비즈니스 로직을 UserService 클래스로 옮겨서 비즈니스 로직을 Service클래스에 집중시켰다.
SessionUtil클래스는 Seesion의 key 값만 Static으로 가지고 있어서 SessionKeys로 이름을 바꿨고 세션에 로그인 아이디 등록하는 메소드는 로그인하는 목적에 맞게 login service로 옮겼다.
UserController에 있던 static LoginRequest, LoginResponse 객체는 코드 가독성을 높이고 일관된 코드를 가지게 하기 위해 따로 dto클래스로 만들었다.
login 컨트롤러의 반환값은 로그인 내용을 보내기에 충분한 LoginResponse의 객체로 바꿨다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/controller/UserController.java:
```diff
@@ -39,64 +33,16 @@ public void registerUser(
-
-        return responseEntity;
-    }
-    //------------------------ request 객체 ---------------------------
```
아 여기 클래스들을 지우라는말이 아니라 주석만 지우라는 얘기였습니다.
여기는 이렇게 inner class로 있어도 되는데 왜 그런지는 고민을 해보시면 좋을 것 같네요

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/UserServiceImpl.java:
```diff
@@ -19,12 +23,30 @@
 
     @Override
-    public UserDto loginUser(String email, String password) {
+    public LoginResponse loginUser(String email, String password, HttpSession session) {
```
이 메소드는 `유저의 정보`에 대한 책임이라기보단 `인증`에 대한 책임으로 보이는데 `UserService`보다는 `LoginService` 같은 클래스를 분리하는게 낫지 않을까요?

hamryt replied:
넵 service와 controller 둘다 분리하도록 하겠습니다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/UserServiceImpl.java:
```diff
@@ -19,12 +23,30 @@
+        LoginResponse loginResponse;
+
+        if (userInfo == null) {
+            loginResponse = LoginResponse.FAIL;
```
로그인에 실패하면 이렇게 리턴값으로 구분하기보다는 예외를 던지는게 좋습니다~
이렇게 리턴을 할 때랑, 예외를 던질때랑 Controller의 코드가 어떻게 달라지는지 비교해보시면 좋을 것 같네요

hamryt replied:
그냥 에러로 처리하지 않고 리턴값으로 구분하면 응답은 항상 정상 으로 나와서 문제가 생긴건지 아닌지를 쉽게 구분하지 못할거 같습니다. 그래서 그것마저 제가 따로 설정을 해야하는 상황이 올거 같습니다. 그래서 로그인 실패시에는 그냥 예외를 받아서 처리하는게 더 나은 방식이라 생각됩니다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/UserServiceImpl.java:
```diff
@@ -19,12 +23,30 @@
+            setLoginUserEmail(session, email);
+
+        } else {
+            log.error("Login Error");
```
이 로그만 보고는 에러의 원인을 디버깅 할 수 있을까요?
본질적으로 봤을 때 error 로그를 찍는 이유는 무엇일까요?

hamryt replied:
예외를 받았을 때 어떤게 에러를 유발하는지 알아보기 쉽게 표시하는 것이라 생각합니다. 에러의 요소를 좀 더 특정할 수 있는것으로 바꾸겠습니다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/UserServiceImpl.java:
```diff
@@ -19,12 +23,30 @@
+
+        } else {
+            log.error("Login Error");
+            throw new RuntimeException("Login Error");
```
여기 또한 어떤 원인으로 이 예외를 던지게 되었는지 더 상세하게 적어주는게 좋습니다

hamryt replied:
넵 알겠습니다.

hamryt committed:
로그인 기능 구현 코드 재배치

Login은 인증에 대한 책임을 가지므로 유저 정보에 대한 책임을 가진 UserController와 UserService로 부터 분리하고 LoginController와 UserService를 만들었습니다.
LoginRequest, LoginResponse는 로그인에서만 사용하는 클래스이므로 LoginController의 내부 클래스로 만들었습니다.
LoginService안의 로그인 실패 처리에 대해선 실패해도 200 상태 코드를 보내지 않기 위해 UserNotExistedException 예외 클래스를 따로 만들어서 처리하였습니다.

hamryt committed:
LoginServiceImpl 안에 UserService로직 분리

email과 password로 해당 유저가 있는지 찾는 것은 loginService의 책임이 아니고 유저의 정보 이므로 userService의 책임인 것으로 판단하여 분리하였습니다.

hamryt committed:
변화 없는 Read query에 일관된 @Transactional(readOnly = ture) 부착

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/controller/LoginController.java:
```diff
@@ -0,0 +1,61 @@
+        @NonNull
+        private String email;
+
+        @NonNull
```
👍

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/controller/UserController.java:
```diff
@@ -24,25 +19,11 @@
     private final UserService userService;
 
     @PostMapping
-    @ResponseStatus(HttpStatus.CREATED)
```
이거는 지워주신 이유가 있으신가요? 201을 리턴하려면 필요해보입니다~

hamryt replied:
아 넵 이유를 모르고 썻었는데 추가하겠습니다.

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/LoginService.java:
```diff
@@ -0,0 +1,8 @@
+import com.hamryt.helparty.dto.UserDto;
+import javax.servlet.http.HttpSession;
+
+public interface LoginService {
```
클래스 선언 아래에 개행을 하나 넣어주면 가독성이 조금 향상됩니다~

hamryt replied:
넵 알겠습니다!

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/LoginServiceImpl.java:
```diff
@@ -0,0 +1,37 @@
+    private final UserMapper userMapper;
+
+    @Transactional
+    public UserDto login(String email, String password, HttpSession session) {
```
`HttpSession`은 메소드로 주입받지 않아도 `@Autowired`로 주입받을 수 있습니다~

hamryt replied:
넵 알겠습니다!

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/LoginServiceImpl.java:
```diff
@@ -0,0 +1,37 @@
+            throw new UserNotExistedException(email);
+        }
+
+        return userMapper.findUserByEmail(email);
```
이 메소드에서는 유저의 정보를 검색하는 쿼리를 2번날리게 되네요. 쿼리를 날리는 것은 무거운 행위이므로 한번만 날리도록 수정해볼 수 있을 것 같습니다~

f-lab-dev commented on helparty/src/main/resources/mybatis/mapper/UserMapper.xml:
```diff
@@ -13,12 +13,12 @@
-    WHERE email = #{email}
-    AND password = #{password}
-    AND status != 'DELETE'
+  <select id="findByEmailAndPassword" resultType="_boolean">
```
이름은 find인데 쿼리 내용을 보면 데이터가 있는지 없는지 확인하는 쿼리네요~
역할에 맞는 이름으로 고쳐주세요~

hamryt replied:
넵!

hamryt committed:
LoginServiceImpl클래스 안에 UserDto객체 조회 쿼리로 대신할 수 있는 존재 여부 확인 쿼리 삭제

f-lab-dev commented on helparty/src/main/java/com/hamryt/helparty/service/LoginServiceImpl.java:
```diff
@@ -14,25 +13,22 @@
-        if (userService.findByEmailAndPassword(email, encryptPassword)) {
-            setLoginUserEmail(session, email);
-        } else {
+        if(userDto!=null){
+            session.setAttribute(SessionKeys.LOGIN_USER_EMAIL, userDto.getEmail());
+        }else{
```
이런 띄어쓰기 컨벤션도 계속 주의해주세요~ checkstyle도 적용해봐주시고요~

```suggestion
        if (userDto!=null){
            session.setAttribute(SessionKeys.LOGIN_USER_EMAIL, userDto.getEmail());
        } else {
```

hamryt replied:
넵 주의하겠습니다.

hamryt committed:
Update helparty/src/main/java/com/hamryt/helparty/service/LoginServiceImpl.java

Co-authored-by: f-lab <54677861+f-lab-dev@users.noreply.github.com>
